### PR TITLE
Add support for .mintfile (#2)

### PR DESCRIPTION
**Pull request by Lutzifer**

### CODE REVIEW

**Comment by LinusU** [comment]
Nice!

Personally, I think that `Mintfile` would be more in line with other Swift/iOS-tools (e.g. `Cartfile`, `Package.swift`)

**Comment by Lutzifer** [comment]
was accidentally thinking more of .ruby-version than Gemfile.

Mintfile it is

**Comment by Lutzifer** [comment]
will change filename and have another look at the tests tomorrow

**Comment by LinusU** [comment]
Cool 😎 

Don't think of me as any authority btw. I'm just a happy user wenting my thoughts 😆

**Comment by Lutzifer** [comment]
Decided, that testing from string is enough

File: Sources/MintKit/Mintfile.swift [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+//  MintKit
+//
+//  Created by Wolfgang Lutz on 04.04.18.
+//
+
+import Foundation
+import PathKit
+
+struct Mintfile {
+  static let defaultPath = Path(FileManager.default.currentDirectoryPath) + "Mintfile"
```
Comment (yonaskolb) [inline Sources/MintKit/Mintfile.swift]: `Path` defaults to the current path, so you can just write `Path("MintFile")

File: Sources/MintKit/Mintfile.swift [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  
+  static func `default`() -> Mintfile? {
+    return self.init(path: Mintfile.defaultPath)
+  }
+  
+  let packageInfos: [PackageInfo]
+  
+  public func version(for repo: String) -> String {
+    return packageInfos.first { $0.repo == repo }?.version ?? ""
```
Comment (yonaskolb) [inline Sources/MintKit/Mintfile.swift]: Maybe we can make this return an optional?

File: Sources/MintKit/Mintfile.swift [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  }
+  
+  init?(path: Path) {
+    guard FileManager.default.fileExists(atPath: path.string) else {
```
Comment (yonaskolb) [inline Sources/MintKit/Mintfile.swift]: can be `path.exists`

File: Sources/MintKit/Mintfile.swift [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+      return nil
+    }
+    
+    guard let contents = try? String(contentsOfFile: path.string, encoding: .utf8) else {
```
Comment (yonaskolb) [inline Sources/MintKit/Mintfile.swift]: You can use `path.read()`

File: Sources/MintKit/Mintfile.swift [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+      fatalError("Could not read mintfile at \(path).")
+    }
+    
+    self.init(string: contents)
+  }
+  
+  init?(string: String) {
+    guard !string.isEmpty else {
```
Comment (yonaskolb) [inline Sources/MintKit/Mintfile.swift]: We can't the file be empty? We don't have to return nil here

File: Sources/MintKit/Mintfile.swift [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    self.packageInfos = linesWithoutTrailingComments
+      .map{$0.trimmingCharacters(in: .whitespacesAndNewlines)}
+      .map { PackageInfo(package: String($0)) }
+    
+    // Print warning for empty version
+    packageInfos
+      .filter { $0.version.isEmpty }
+      .forEach { print("🌱  MINTFILE: repository \($0.repo) has no defined version. Specify a version using <Repo>@<Commitish>.") }
+
+    // Print warning for multiple definitions
```
Comment (yonaskolb) [inline Sources/MintKit/Mintfile.swift]: I've become a big fan of the new Dictionary initializers in Swift 4.

This gives you a dictionary of [duplicatedRepos: versions]
```swift
let duplicates = Dictionary(grouping: packageInfos, by: { $0.repo })
  .filter { $0.value.count > 0 }
  .mapValues { $0.map { $0.version } }
```

```swift
duplicates.forEach { repo, versions in
  print("🌱  MINTFILE: repository \"\(repo)\" defined multiple times, using versions \"\(versions.joined(separator: ", "))\".")
}
```

File: Sources/MintKit/Mintfile.swift [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+import PathKit
+
+struct Mintfile {
+  static let defaultPath = Path(FileManager.default.currentDirectoryPath) + "Mintfile"
+  
+  static func `default`() -> Mintfile? {
+    return self.init(path: Mintfile.defaultPath)
+  }
+  
+  let packageInfos: [PackageInfo]
```
Comment (yonaskolb) [inline Sources/MintKit/Mintfile.swift]: The struct `PackageInfo` has always had a weird name. I think I'll change it to `Package`.
Can we rename this property to `packages`.

**Comment by Lutzifer** [comment]
@yonaskolb nice, need to have  a look into pathkit sometime

Did the other changes also.

We could somehow merge MintPackage (fka PackageInfo) and Package i guess.

File: Sources/MintKit/Mintfile.swift (hunk 6 above)
Comment (Lutzifer) [reply Sources/MintKit/Mintfile.swift]: after changing $0.value.count > 0 to $0.value.count > 1 this really worked nicely

**Comment by yonaskolb** [comment]
Thanks @Lutzifer. Yeah I'm probably going to be doing a bunch of refactoring in Mint, and get it away from it's prototype roots. I'll look at the Package structs then as well 👍 

In regards to this PR, we could merge it now and then we can add `mint bootstrap/install` in a later PR.
Before we do that though, what do you guys think of the name `MintPackages` or `Mint.packages` for the file? I like `Mint` as well, but it might conflict if someone had the `mint` executable in their project directory

**Comment by Lutzifer** [comment]
Cartfile, Podfile, Gemfile, Fastfile, Mintfile. I think we are in good company with the name 😉

**Comment by yonaskolb** [comment]
True, they’re some good examples 😀👍

**Pull request merged**